File: backend/app/execution/capabilities/reorder_rows.py
"""reorderRows — rearrange rows in place: moveMatching / reverse / clusterByKey."""

from __future__ import annotations

from typing import Any, Optional

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range
from app.execution.utils import normalize_for_compare, parse_number_flexible


def _ensure_2d(values: Any) -> list[list[Any]]:
    if values is None:
        return []
    if not isinstance(values, list):
        return [[values]]
    if not values:
        return []
    if not isinstance(values[0], list):
        return [values] if len(values) > 1 else [[values[0]]]
    return values


def _test_condition(cell: Any, condition: str, value: Any) -> bool:
    # Normalize strings for equality/contains so Hebrew + RTL marks + NBSP
    # all compare consistently. Use parse_number_flexible for numeric
    # comparisons so text-stored numbers work.
    if condition == "blank":
        return normalize_for_compare(cell) == ""
    if condition == "notBlank":
        return normalize_for_compare(cell) != ""
    if condition == "equals":
        if isinstance(cell, (int, float)) and isinstance(value, (int, float)):
            return cell == value
        return normalize_for_compare(cell) == normalize_for_compare(value)
    if condition == "notEquals":
        if isinstance(cell, (int, float)) and isinstance(value, (int, float)):
            return cell != value
        return normalize_for_compare(cell) != normalize_for_compare(value)
    if condition == "contains":
        return normalize_for_compare(value) in normalize_for_compare(cell)
    if condition == "notContains":
        return normalize_for_compare(value) not in normalize_for_compare(cell)
    if condition == "greaterThan":
        n = parse_number_flexible(cell)
        vv = parse_number_flexible(value)
        return n is not None and vv is not None and n > vv
    if condition == "lessThan":
        n = parse_number_flexible(cell)
        vv = parse_number_flexible(value)
        return n is not None and vv is not None and n < vv
    return False


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    address = params.get("range")
    mode = params.get("mode")
    condition_col = params.get("conditionColumn")
    condition = params.get("condition")
    condition_value = params.get("conditionValue")
    destination = params.get("destination", "top")
    has_headers = bool(params.get("hasHeaders", True))

    if not address or not mode:
        return {"status": "error", "message": "reorderRows requires 'range' and 'mode'."}

    if mode in ("moveMatching", "clusterByKey") and condition_col is None:
        return {"status": "error", "message": f"mode={mode} requires 'conditionColumn'."}

    rng = resolve_range(ctx.workbook_handle, address)
    used = rng.current_region if rng.count > 1 else rng
    raw = _ensure_2d(used.value)
    if not raw:
        return {"status": "success", "message": "Range is empty."}

    header_row: Optional[list[Any]] = raw[0] if has_headers else None
    data_rows = [list(r) for r in (raw[1:] if has_headers else raw)]

    moved_count = 0
    if mode == "reverse":
        new_order = list(reversed(data_rows))
    elif mode == "moveMatching":
        matching: list[list[Any]] = []
        rest: list[list[Any]] = []
        for row in data_rows:
            if _test_condition(row[condition_col] if len(row) > condition_col else None, condition or "", condition_value):
                matching.append(row)
            else:
                rest.append(row)
        moved_count = len(matching)
        new_order = (matching + rest) if destination == "top" else (rest + matching)
    elif mode == "clusterByKey":
        buckets: dict[str, list[list[Any]]] = {}
        key_order: list[str] = []
        for row in data_rows:
            raw_key = row[condition_col] if len(row) > condition_col else None
            # Normalize keys so Hebrew with RTL marks, NBSP padding, or NFC
            # variants cluster together.
            key = normalize_for_compare(raw_key)
            if key not in buckets:
                buckets[key] = []
                key_order.append(key)
            buckets[key].append(row)
        new_order = [r for k in key_order for r in buckets[k]]
        moved_count = sum(len(b) for b in buckets.values() if len(b) > 1)
    else:
        return {"status": "error", "message": f"Unknown mode: {mode}"}

    output = ([header_row] if header_row else []) + new_order

    if ctx.dry_run:
        return {"status": "preview", "message": f"Would reorderRows mode={mode} on {used.address}."}

    sheet = used.sheet
    top_left = used[0, 0]
    try:
        target = sheet.range(
            (top_left.row, top_left.column),
            (top_left.row + len(output) - 1, top_left.column + len(output[0]) - 1),
        )
        target.value = output
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"Write failed: {exc}", "error": str(exc)}

    if mode == "reverse":
        msg = f"Reversed {len(data_rows)} rows in {used.address}."
    elif mode == "moveMatching":
        msg = f"Moved {moved_count} matching row(s) to {destination} of {used.address}."
    else:
        msg = f"Clustered rows in {used.address} by column {condition_col} ({moved_count} row(s) in multi-row groups)."

    return {
        "status": "success",
        "message": msg,
        "outputs": {"range": used.address, "movedRowCount": moved_count},
    }


registry.register("reorderRows", handler, mutates=True)

File: frontend/src/engine/capabilities/reorderRows.ts
/**
 * reorderRows — rearrange rows in-place by criteria.
 *
 * Modes:
 *  - "moveMatching": rows whose conditionColumn satisfies condition go to
 *    destination ("top" or "bottom" of the range). Order within the moved
 *    group and within the kept group is preserved.
 *  - "reverse": flip row order in the range. Headers are respected.
 *  - "clusterByKey": group rows sharing the same value in conditionColumn.
 *    First-appearance order of each key is preserved, but all rows with
 *    a given key become contiguous.
 *
 * Single context.sync — reads values, computes new order in JS, writes back.
 */

import { CapabilityMeta, ReorderRowsParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";
import { normalizeForCompare } from "../utils/normalizeString";
import { parseNumberFlexible } from "../utils/parseNumberFlexible";
import { ensureUnmerged } from "../utils/mergedCells";

const meta: CapabilityMeta = {
  action: "reorderRows",
  description:
    "Rearrange rows in-place: move matching rows, reverse order, or cluster by key",
  mutates: true,
  affectsFormatting: false,
};

type Row = (string | number | boolean | null)[];

function testCondition(
  cell: unknown,
  condition: ReorderRowsParams["condition"],
  value: unknown,
): boolean {
  // Normalize both sides of string comparisons so Hebrew with trailing RTL
  // marks, NBSP, or stray whitespace still matches. Use parseNumberFlexible
  // for numeric comparisons so text-stored numbers ("1,234", "$100") work.
  switch (condition) {
    case "blank":    return normalizeForCompare(cell) === "";
    case "notBlank": return normalizeForCompare(cell) !== "";
    case "equals": {
      if (typeof cell === "number" && typeof value === "number") return cell === value;
      return normalizeForCompare(cell) === normalizeForCompare(value);
    }
    case "notEquals": {
      if (typeof cell === "number" && typeof value === "number") return cell !== value;
      return normalizeForCompare(cell) !== normalizeForCompare(value);
    }
    case "contains":
      return normalizeForCompare(cell).includes(normalizeForCompare(value));
    case "notContains":
      return !normalizeForCompare(cell).includes(normalizeForCompare(value));
    case "greaterThan": {
      const n = parseNumberFlexible(cell);
      const v = parseNumberFlexible(value);
      return n !== null && v !== null && n > v;
    }
    case "lessThan": {
      const n = parseNumberFlexible(cell);
      const v = parseNumberFlexible(value);
      return n !== null && v !== null && n < v;
    }
    default: return false;
  }
}

async function handler(
  context: Excel.RequestContext,
  params: ReorderRowsParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { range: address, mode, conditionColumn, condition, conditionValue, destination = "top", hasHeaders = true } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would reorderRows mode=${mode} on ${address}` };
  }

  options.onProgress?.(`Reordering rows (${mode})...`);

  const rawRange = resolveRange(context, address);
  const used = rawRange.getUsedRange(false);
  used.load(["values", "rowCount", "columnCount", "address"]);
  await context.sync();

  // Merged cells make reordering unsafe — non-anchor cells come through as
  // null, and the write-back would produce misaligned data. Auto-unmerge.
  const mergeReport = await ensureUnmerged(context, used, {
    operation: "reorderRows",
    policy: "refuseWithError",
  });
  if (mergeReport.error) return mergeReport.error;
  if (mergeReport.hadMerges) {
    used.load(["values"]);
    await context.sync();
  }

  const values = (used.values ?? []) as Row[];
  if (values.length === 0) {
    return { stepId: "", status: "success", message: "Range is empty." };
  }

  const headerRow = hasHeaders ? values[0] : null;
  const dataRows = hasHeaders ? values.slice(1) : values.slice();

  let newOrder: Row[];
  let movedCount = 0;

  switch (mode) {
    case "reverse": {
      newOrder = dataRows.slice().reverse();
      break;
    }
    case "moveMatching": {
      if (conditionColumn === undefined) {
        return { stepId: "", status: "error", message: "moveMatching requires conditionColumn." };
      }
      const matching: Row[] = [];
      const rest: Row[] = [];
      for (const row of dataRows) {
        if (testCondition(row[conditionColumn], condition, conditionValue)) matching.push(row);
        else rest.push(row);
      }
      movedCount = matching.length;
      newOrder = destination === "top" ? [...matching, ...rest] : [...rest, ...matching];
      break;
    }
    case "clusterByKey": {
      if (conditionColumn === undefined) {
        return { stepId: "", status: "error", message: "clusterByKey requires conditionColumn." };
      }
      // Preserve first-appearance order of each unique key. Keys are
      // normalized for bucketing so Hebrew with trailing RTL marks,
      // NBSP-padded, or NFC-variant strings cluster together.
      const buckets = new Map<string, Row[]>();
      const keyOrder: string[] = [];
      for (const row of dataRows) {
        const key = normalizeForCompare(row[conditionColumn]);
        if (!buckets.has(key)) { buckets.set(key, []); keyOrder.push(key); }
        buckets.get(key)!.push(row);
      }
      newOrder = [];
      for (const key of keyOrder) for (const r of buckets.get(key)!) newOrder.push(r);
      movedCount = Array.from(buckets.values()).filter((arr) => arr.length > 1)
        .reduce((sum, arr) => sum + arr.length, 0);
      break;
    }
    default:
      return { stepId: "", status: "error", message: `Unknown mode: ${mode}` };
  }

  // Compose output and write back.
  const output: Row[] = headerRow ? [headerRow, ...newOrder] : newOrder;

  // Anchor write at used-range top-left.
  const sheet = used.worksheet;
  const addrPart = used.address.includes("!") ? used.address.split("!").pop()! : used.address;
  const [topLeftRef] = addrPart.split(":");
  const m = topLeftRef.match(/^([A-Z]+)(\d+)$/);
  if (!m) return { stepId: "", status: "error", message: `Could not parse used range: ${used.address}` };
  let startCol = 0;
  for (let c = 0; c < m[1].length; c++) startCol = startCol * 26 + (m[1].charCodeAt(c) - 64);
  startCol -= 1;
  const startRow = Number(m[2]) - 1;

  try {
    const out = sheet.getRangeByIndexes(startRow, startCol, output.length, used.columnCount);
    out.values = output as unknown as (string | number | boolean)[][];
    await context.sync();
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    return { stepId: "", status: "error", message: `Failed to write reordered rows: ${msg}`, error: msg };
  }

  const baseMessage =
    mode === "reverse"
      ? `Reversed ${dataRows.length} rows in ${address}.`
      : mode === "moveMatching"
      ? `Moved ${movedCount} matching row(s) to the ${destination} of ${address}.`
      : `Clustered rows in ${address} by column ${conditionColumn} (${movedCount} row(s) in multi-row groups).`;

  return {
    stepId: "",
    status: "success",
    message: `${baseMessage}${mergeReport.warning ?? ""}`,
    outputs: { range: used.address, movedRowCount: movedCount },
  };
}

registry.register(meta, handler as any);
export { meta };
